File: leetcode/solutions/0278.FirstBadVersion/firstBadVersion.ts
export type IsBadVersion = (n: number) => boolean;
export type FindFirstBadVirsion = (n: number) => number;

export default function solution(
  isBadVersion: IsBadVersion
): FindFirstBadVirsion {
  return function(n: number): number {
    let l = 1;
    let r = n;

    while (l < r) {
      const mid = (l + r) >>> 1;

      if (isBadVersion(mid)) {
        r = mid;
      } else {
        l = mid + 1;
      }
    }

    return l;
  };
}

File: leetcode/solutions/0278.FirstBadVersion/firstBadVersion.go
package solution

func firstBadVersion(n int) int {
	l, r := 0, n

	for l < r {
		mid := (l + r) / 2

		if isBadVersion(mid) {
			r = mid
		} else {
			l = mid + 1
		}
	}

	return l
}

// FirstBadVersion represents the first bad version that is the answer.
var FirstBadVersion int

func isBadVersion(i int) bool {
	return FirstBadVersion <= i
}
